src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
```

Comment:
Then we should never block reading if there's no data available to fill the whole buffer but return with the amount we do have (except block if none) like I've agreed to refactor the reading to include a while loop again.

Though, what's the verdict on the final design of the stream? One thing I did realize is that if you pass the output stream to input stream constructor, they would need to have some shared private API to pass the data which shouldn't be exposed even internally to any other part of SSH.NET. Current implementation avoids that by having a shared data structure - the queue.

This would be all much easier to discuss in a closed room with a whiteboard and beverages.